ding/utils/data/dataset.py

```diff
@@ -69,6 +70,43 @@ def _load_d4rl(self, dataset: Dict[str, np.ndarray]) -> None:
             self._data.append(trans_data)
 
 
+@DATASET_REGISTRY.register('hdf5')
+class HDF5Dataset(Dataset):
+
+    def __init__(self, cfg: dict) -> None:
+        try:
+            import h5py
+        except ImportError:
+            logging.warning("not found h5py env, please install it")
+        data_path = cfg.policy.learn.get('data_path', None)
+        data = h5py.File(data_path, 'r')
+        self._load_data(data)
+
+    def __len__(self) -> int:
+        return len(self._data['obs'])
+
+    def __getitem__(self, idx: int) -> Dict[str, torch.Tensor]:
+        return {k: self._data[k][idx] for k in self._data.keys()}
+
+    def _load_data(self, dataset: Dict[str, np.ndarray]) -> None:
+        self._data = {}
+        for k in dataset.keys():
+            logging.info(f'Load {k} data.')
+            self._data[k] = dataset[k][:]
+
+
+def hdf5_save(exp_data, expert_data_path):
+    import h5py
+    import numpy as np
+    dataset = dataset = h5py.File('%s_demos.hdf5' % expert_data_path.replace('.pkl', ''), 'w')
+    dataset.create_dataset('obs', data=np.array([d['obs'].numpy() for d in exp_data]), compression='gzip')
```

Comment:
Can we implement this function more generally? For example, use `to_ndarray` function to transform into numpy type, and traverse all the keys in element0 (assume all the elements have the same format) to call `dataset.create_dataset`.

Reply:
We plan to implement in the future due to the following points:
1. The saved in the buffer transitions include different key-value pairs depending on RL algorithms. For example, ppo saves extra `logit(\mu, \sigma)` in transition.
2. Taking into the above extra key-value pairs will cost more memory and time.